pkg/fs/fs_test.go

```diff
@@ -0,0 +1,160 @@
+		} else if got, exp := newContents, newpath; got != exp {
+			t.Fatalf("got contents %q, expected %q", got, exp)
+		}
+
+		if err := fs.RenameFileWithReplacement(oldpath, newpath); err != nil {
+			t.Fatalf("ReplaceFileIfExists returned an error: %s", err)
+		}
+
+		if err := fs.SyncDir(filepath.Dir(oldpath)); err != nil {
+			panic(err)
+		}
+
+		// Contents of newpath will now be equivalent to oldpath' contents.
+		newContents = MustReadAllFile(newpath)
+		if newContents != oldContents {
+			t.Fatalf("contents for files differ: %q versus %q", newContents, oldContents)
+		}
+
+		// oldpath will be removed.
+		if MustFileExists(oldpath) {
+			t.Fatalf("file %q still exists, but it shouldn't", oldpath)
+		}
+	})
+
+	t.Run("not exists", func(t *testing.T) {
+		oldpath := MustCreateTempFile()
+		defer MustRemoveAll(oldpath)
+
+		oldContents := MustReadAllFile(oldpath)
+		if got, exp := oldContents, oldpath; got != exp {
+			t.Fatalf("got contents %q, expected %q", got, exp)
+		}
+
+		root := filepath.Dir(oldpath)
+		newpath := filepath.Join(root, "foo")
+		if err := fs.RenameFileWithReplacement(oldpath, newpath); err != nil {
+			t.Fatalf("ReplaceFileIfExists returned an error: %s", err)
+		}
+
+		if err := fs.SyncDir(filepath.Dir(oldpath)); err != nil {
+			panic(err)
+		}
+
+		// Contents of newpath will now be equivalent to oldpath's contents.
+		newContents := MustReadAllFile(newpath)
+		if newContents != oldContents {
+			t.Fatalf("contents for files differ: %q versus %q", newContents, oldContents)
+		}
+
+		// oldpath will be removed.
+		if MustFileExists(oldpath) {
+			t.Fatalf("file %q still exists, but it shouldn't", oldpath)
+		}
+	})
+}
+
+func TestCreateFileWithReplacement(t *testing.T) {
+	path := MustCreateTempFile()
+	defer MustRemoveAll(path)
+
+	// should return an error if we CreateFile to the same path
+	_, err := fs.CreateFile(path)
+	if err == nil {
+		t.Fatalf("CreateFile did not return an error")
+	}
+
+	// contents of the file should be intact
+	contents := MustReadAllFile(path)
+	if got, exp := contents, path; got != exp {
+		t.Fatalf("got contents %q, expected %q", got, exp)
+	}
+
+	// running CreateFileWithReplacement on path should not return an error
+	_, err = fs.CreateFileWithReplacement(path)
+	if err != nil {
+		t.Fatalf("CreateFileWithReplacement returned err: %v", err)
+	}
+
+	// the file at path should now be empty
+	contents = MustReadAllFile(path)
+	if contents != "" {
+		t.Fatalf("expected file to be empty but got: %v", contents)
+	}
+
+}
+
+// MustCreateTempFile creates a temporary file returning the path to the file.
+//
+// MustCreateTempFile writes the absolute path to the file into the file itself.
+// It panics if there is an error.
+func MustCreateTempFile() string {
+	f, err := ioutil.TempFile("", "fs-test")
+	if err != nil {
+		panic(fmt.Sprintf("failed to create temp file: %v", err))
+	}
+
+	name := f.Name()
+	f.WriteString(name)
+	if err := f.Close(); err != nil {
+		panic(err)
+	}
+	return name
+}
```

Comment:
Minor nit but it's not clear from the name that it's writing its own name to the filename. I think it would be clearer from the caller to pass in a string for the `data` to write. Also, probably worth adding an error check on `os.File.WriteString()`:

```go
// MustCreateTempFile creates a temporary file returning the path to the file.
func MustCreateTempFile(data string) string {
	f, err := ioutil.TempFile("", "fs-test")
	if err != nil {
		panic(fmt.Sprintf("failed to create temp file: %v", err))
	} else if _, err := f.WriteString(data); err != nil {
		panic(err)
	} else if err := f.Close(); err != nil {
		panic(err)
	}
	return name
}
```

Another approach I've been doing more lately is making the function a helper and naming it something like `CreateTempFileOrFail()`. That way it only fails the test instead of the whole program execution (although temp files are unlikely to fail).

```go
// CreateTempFileOrFail creates a temporary file returning the path to the file.
func MustCreateTempFile(t testing.TB, data string) string {
	tb.Helper()
	
	f, err := ioutil.TempFile("", "fs-test")
	if err != nil {
		tb.Fatalf("failed to create temp file: %v", err)
	} else if _, err := f.WriteString(data); err != nil {
		t.Fatal(err)
	} else if err := f.Close(); err != nil {
		t.Fatal(err)
	}
	return name
}
```